### PR TITLE
feat(`api/rpc`): OpenRPC scaffolding, migrating OpenAPI gateway to `api/gateway`

### PR DESCRIPTION
**Pull request by distractedm1nd**
Based on #1056

- Moves the gateway to `api/gateway`. This is no longer bundled with the node as of right now.
- Removes `/service` directory
- Creates new JSON-RPC server in `api/rpc`
- Moves rpc Config definition to `nodebuilder/rpc` (no more cyclical dependency problem)

I would suggest reintroducing the gateway, with an additional config and a default "off" mode, should be done in a different PR.

### CODE REVIEW

File: api/rpc/server.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/filecoin-project/go-jsonrpc"
+)
+
+type Server struct {
+	srv      *http.Server
+	handler  *jsonrpc.RPCServer
```
Comment (renaynay) [inline api/rpc/server.go:14]: this naming is confusing, maybe lets do 

`http` and `rpc` ?

File: api/rpc/server.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		srv: &http.Server{
+			Addr:    address + ":" + port,
+			Handler: handler,
+			// the amount of time allowed to read request headers. set to the default 2 seconds
+			ReadHeaderTimeout: 2 * time.Second,
+		},
+	}
+}
+
+func (s *Server) Start(context.Context) error {
```
Comment (renaynay) [inline api/rpc/server.go:31]: doc

File: nodebuilder/config.go [hunk 3]
```diff
@@ -4,6 +4,8 @@ import (
 	"io"
 	"os"
 
+	"github.com/celestiaorg/celestia-node/nodebuilder/rpc"
```
Comment (renaynay) [inline nodebuilder/config.go:7]: format

File: nodebuilder/node.go [hunk 4]
```diff
@@ -6,6 +6,8 @@ import (
 	"strings"
 	"time"
 
+	"github.com/celestiaorg/celestia-node/api/rpc"
```
Comment (renaynay) [inline nodebuilder/node.go:9]: format

File: nodebuilder/rpc/module.go [hunk 5]
```diff
@@ -3,28 +3,29 @@ package rpc
 import (
 	"context"
 
+	"github.com/celestiaorg/celestia-node/api/rpc"
```
Comment (renaynay) [inline nodebuilder/rpc/module.go:6]: format

File: api/rpc/handler.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	return &Handler{
+		state:  state,
+		share:  share,
+		header: header,
+		das:    das,
+	}
+}
+
+func (h *Handler) RegisterEndpoints(rpc *Server) {
+	rpc.handler.Register("State", h.state)
```
Comment (renaynay) [inline api/rpc/handler.go:36]: Does namespace have to be capitalised?

**Review by renaynay — COMMENTED** [review]
needs some godoc and format but otherwise looks good already.

File: api/rpc/handler.go (hunk 6 above)
Comment (distractedm1nd) [reply api/rpc/handler.go:36]: no, but I did find out that they need to all be the same in the next PR - its only one connection per namespace

File: nodebuilder/rpc/module.go (hunk 5 above)
Comment (distractedm1nd) [reply nodebuilder/rpc/module.go]: done

File: nodebuilder/node.go (hunk 4 above)
Comment (distractedm1nd) [reply nodebuilder/node.go]: done

File: nodebuilder/config.go (hunk 3 above)
Comment (distractedm1nd) [reply nodebuilder/config.go]: done

File: api/rpc/server.go (hunk 2 above)
Comment (distractedm1nd) [reply api/rpc/server.go]: done

File: api/rpc/server.go (hunk 1 above)
Comment (distractedm1nd) [reply api/rpc/server.go]: done

File: api/rpc/server.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// Start starts the RPC Server.
+func (s *Server) Start(context.Context) error {
+	listener, err := net.Listen("tcp", s.http.Addr)
+	if err != nil {
+		return err
+	}
+	s.listener = listener
+	log.Infow("RPC server started", "listening on", listener.Addr().String())
+	//nolint:errcheck
+	go s.http.Serve(listener)
```
Comment (walldiss) [inline api/rpc/server.go:40]: Looks like we are listening for just `tcp` here anyway. Maybe we can replace explicit listener here with  more convinient `ListenAndServe` and also don't store it in  the `Server` struct?

**Review by Wondertan — CHANGES_REQUESTED** [review]
RPC is not a module and should not have `nodebuilder`. It's like commands, config, or flags of each module. Each module has an API and each module can be accessed through RPC.

If the RPC server is provided to DI, each module should register its handler there, not some high-level module or API importing it and registering handlers manually. This is something that should be done in this or other RPC related PR and not deferred to some later refactorings.

Gateway though could be a module.

**Comment by distractedm1nd** [comment]
I disagree here. The RPC is a service that is running for the lifecycle of the node, and accesses the other modules - just like the other services. I do not understand the comparison with commands, config, or flags.

Generally, by letting each module be in charge of registering itself on the RPC server, it will no longer be possible to see at a glance which APIs the RPC is exposing. More importantly, this clearly leads to higher coupling (of the individual modules to the RPC) and lower cohesion (inside the individual nodebuilder modules, since they now have to register on an RPC). I can explain this and the consequences more in depth if you disagree.

**Comment by renaynay** [comment]
> RPC is not a module and should not have `nodebuilder`. It's like commands, config, or flags of each module. Each module has an API and each module can be accessed through RPC.
> 
> If the RPC server is provided to DI, each module should register its handler there, not some high-level module or API importing it and registering handlers manually. This is something that should be done in this or other RPC related PR and not deferred to some later refactorings.
> 
> Gateway though could be a module.

@Wondertan I think rpc warrants its own fx.Module -- it is not one of our "Modules" offered by node per-se but I would consider it a module in our software. 

I am, however, okay with each module registering its own endpoint on RPC, but I don't really see the added benefit rn.

File: api/rpc/server.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+type Server struct {
+	http     *http.Server
+	rpc      *jsonrpc.RPCServer
+	listener net.Listener
+}
+
+func NewServer(address string, port string) *Server {
+	handler := jsonrpc.NewServer()
+	return &Server{
+		rpc: handler,
```
[the remark this answers is not in the document]
Comment (distractedm1nd) [reply api/rpc/server.go:21]: I wasn't sure what this was referencing, so I renamed that handler variable to rpc. Or did you mean the Server struct -> RPC?

File: api/rpc/server.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		http: &http.Server{
+			Addr:    address + ":" + port,
+			Handler: handler,
+			// the amount of time allowed to read request headers. set to the default 2 seconds
+			ReadHeaderTimeout: 2 * time.Second,
```
[the remark this answers is not in the document]
Comment (distractedm1nd) [reply api/rpc/server.go:26]: #1276

Not sure what they should be configured to, so I also think its better to be in a different issue

**Review by renaynay — DISMISSED** [review]
* 	// rpc components
	RPCServer *rpc.Server // not optional
*

File: api/rpc/server.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (walldiss) [reply api/rpc/server.go]: You got it right

**Comment by walldiss** [comment]
I think each module should expose / return `ServeHTTP` or other `handler`- type interface so they can be registered manually in nodebuilder or other higher level caller. It will only expose required handler logic without need to pass transport or middleware level of abstractions inside the components that they don't have any use of. So we can construct server as an independent module and assign all required transport settings and middleware according to our vision of node, not of the pkg itself.

**Review by Wondertan — APPROVED** [review]
After plenty of discussions, we(@distractedm1nd, @renaynay, @vgonkivs, and @Wondertan) agreed on this approach.

**Review by vgonkivs — APPROVED** [review]

**Review by renaynay — APPROVED** [review]
lfg

**Pull request merged**